include/revng/Pipeline/Containers.h

```diff
@@ -0,0 +1,227 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstddef>
+#include <fstream>
+#include <memory>
+#include <optional>
+#include <set>
+#include <sstream>
+#include <string>
+
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+#include "revng/Pipeline/ContainerFactory.h"
+#include "revng/Pipeline/InputOutputContract.h"
+#include "revng/Pipeline/Target.h"
+#include "revng/Support/Assert.h"
+
+namespace Pipeline {
+
+/// A wrapper around a optional<container> and the pointer to the factory that
+/// can spawn this container when needed
+class MaybeContainer {
+private:
+  std::optional<std::unique_ptr<ContainerBase>> Container;
+  ContainerFactory *Factory;
+
+public:
+  MaybeContainer(ContainerFactory &Factory,
+                 std::unique_ptr<ContainerBase> Container) :
+    Container(std::move(Container)), Factory(&Factory) {}
+
+  MaybeContainer(ContainerFactory &Factory) :
+    Container(std::nullopt), Factory(&Factory) {}
+
+  ContainerBase &get() { return **Container; }
+
+  const ContainerBase &get() const { return **Container; }
+
+  ContainerBase &getOrCreate() {
+    if (not hasValue())
+      Container = Factory->createEmpty();
+
+    return get();
+  }
+
+  bool hasValue() const { return Container.has_value(); }
+
+  void mergeBack(MaybeContainer &&Other) {
+    if (not Other.hasValue())
+      return;
+
+    if (not hasValue()) {
+      *this = std::move(Other);
+      return;
+    }
+
+    get().mergeBack(std::move(Other.get()));
+  }
+
+  std::unique_ptr<ContainerBase> extract() {
+    if (not hasValue())
+      return nullptr;
+
+    auto ptr = std::move(*Container);
+    Container = std::nullopt;
+    return ptr;
+  }
+
+  MaybeContainer cloneFiltered(const TargetsList &Targets) const {
+    if (not hasValue())
+      return MaybeContainer(*Factory);
+    return MaybeContainer(*Factory, get().cloneFiltered(Targets));
+  }
+
+  llvm::Error storeToDisk(llvm::StringRef Directory) const {
+    if (not hasValue())
+      return llvm::Error::success();
+
+    return get().storeToDisk(Directory);
+  }
+
+  llvm::Error loadFromDisk(llvm::StringRef File) {
+    if (not llvm::sys::fs::exists(File)) {
+      Container = std::nullopt;
+      return llvm::Error::success();
+    }
+
+    return getOrCreate().loadFromDisk(File);
+  }
+
+  TargetsList enumerate() const {
+    if (not hasValue())
+      return {};
+
+    return get().enumerate();
+  }
+};
+
+/// a map from containerName to a MaybeContainer, which is initially empty.
+class Containers {
```

Comment:
Sounds like this should be `ContainersSomething` that you can `.add` to. `ContainersMap`?
It might be a good idea to provide a `std::map`-like API, this way you set an expectation of behavior similar to something you already know and then you make sure you provide it.